pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch

```diff
@@ -12,33 +12,33 @@ tries to gather environments with that they call
 "environment-generators" and then seems to pass that on to all the other
 executables that are being called from managers.
 ---
- src/core/manager.c | 13 +++++++++----
- 1 file changed, 9 insertions(+), 4 deletions(-)
+ src/core/manager.c | 8 +++++++-
+ 1 file changed, 7 insertions(+), 1 deletion(-)
 
 diff --git a/src/core/manager.c b/src/core/manager.c
-index 5b0bdb1bc7..1538a5200a 100644
+index 71ef7f27b4..922d10f1b6 100644
 --- a/src/core/manager.c
 +++ b/src/core/manager.c
-@@ -3653,10 +3653,15 @@ static int manager_run_generators(Manager *m) {
-         argv[4] = NULL;
+@@ -3788,12 +3788,18 @@ static int manager_run_generators(Manager *m) {
+         }
  
          RUN_WITH_UMASK(0022)
--                (void) execute_directories((const char* const*) paths, DEFAULT_TIMEOUT_USEC, NULL, NULL,
--                                           (char**) argv, m->transient_environment,
--                                           EXEC_DIR_PARALLEL | EXEC_DIR_IGNORE_ERRORS | EXEC_DIR_SET_SYSTEMD_EXEC_PID);
--
-+                (void) execute_directories((const char* const*) paths, DEFAULT_TIMEOUT_USEC,
-+                                    // On NixOS we must propagate PATH to generators so they are
-+                                    // able to find binaries such as `fsck.${fstype}` and
-+                                    // `mkfs.${fstype}`. That is why the last argument of the
-+                                    // function (envp) is set to NULL. This propagates systemd's
-+                                    // environment (e.g. PATH) that was setup
-+                                    // before calling systemd from stage-2-init.sh.
-+                                    NULL, NULL, (char**) argv, /* NixOS: use inherited env */ NULL,
-+                                    EXEC_DIR_PARALLEL | EXEC_DIR_IGNORE_ERRORS | EXEC_DIR_SET_SYSTEMD_EXEC_PID);
-         r = 0;
++                // On NixOS we must propagate PATH to generators so they are
++                // able to find binaries such as `fsck.${fstype}` and
++                // `mkfs.${fstype}`. That is why the last argument of the
++                // function (envp) is set to NULL. This propagates systemd's
++                // environment (e.g. PATH) that was setup
++                // before calling systemd from stage-2-init.sh.
+                 (void) execute_directories(
+                                 (const char* const*) paths,
+                                 DEFAULT_TIMEOUT_USEC,
+                                 /* callbacks= */ NULL, /* callback_args= */ NULL,
+                                 (char**) argv,
+-                                ge,
++                                /* NixOS: use inherited env */ NULL,
```

Comment:
Here you're not only discarding environment variables that came from outside, but also ones that are generated by `build_generator_environment` function. Our version of the patch prevents that function from copying environment variables from `transient_environment` instead: https://github.com/tweag/nixpkgs/blob/systemd-boot-zfs/pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch - this should remove all unneeded variables, while keeping essential ones.

Reply:
applied.